tests/php/FilesystemPublisherTest.php

```diff
@@ -0,0 +1,360 @@
+<?php
+
+namespace SilverStripe\StaticPublishQueue\Test;
+
+use SilverStripe\Assets\Filesystem;
+use SilverStripe\CMS\Model\RedirectorPage;
+use SilverStripe\CMS\Model\SiteTree;
+use SilverStripe\Control\Director;
+use SilverStripe\Control\HTTPApplication;
+use SilverStripe\Core\Config\Config;
+use SilverStripe\Core\Environment;
+use SilverStripe\Dev\SapphireTest;
+use SilverStripe\Dev\TestKernel;
+use SilverStripe\StaticPublishQueue\Extension\Publishable\PublishableSiteTree;
+use SilverStripe\StaticPublishQueue\Publisher\FilesystemPublisher;
+use SilverStripe\StaticPublishQueue\Test\StaticPublisherTest\Model\StaticPublisherTestPage;
+use SilverStripe\View\SSViewer;
+use Symbiote\QueuedJobs\Services\QueuedJobService;
+
+/**
+ * Tests for the {@link FilesystemPublisher} class.
+ *
+ * @package staticpublishqueue
+ */
+class FilesystemPublisherTest extends SapphireTest
+{
+    protected $usesDatabase = true;
+
+    /**
+     * @var FilesystemPublisher
+     */
+    private $fsp = null;
+
+    protected static $required_extensions = [
+        SiteTree::class => [
+            PublishableSiteTree::class,
+        ],
+    ];
+
+    protected function setUp()
+    {
+        parent::setUp();
+
+        Config::modify()->set(FilesystemPublisher::class, 'domain_based_caching', false);
+        Config::modify()->set(QueuedJobService::class, 'use_shutdown_function', false);
+        Config::modify()->set(Director::class, 'alternate_base_url', 'http://example.com/');
+
+        $mockFSP = $this->getMockBuilder(FilesystemPublisher::class)->setMethods([
+            'getHTTPApplication',
+        ])->getMock();
+
+        $mockFSP->method('getHTTPApplication')->willReturnCallback(function () {
+            return new HTTPApplication(new TestKernel(BASE_PATH));
+        });
+
+        $this->fsp = $mockFSP->setDestFolder('cache/testing/');
+    }
+
+    protected function tearDown()
+    {
+        if ($this->fsp !== null && file_exists($this->fsp->getDestPath())) {
+            Filesystem::removeFolder($this->fsp->getDestPath());
+        }
+        parent::tearDown();
+    }
+
+    public function testUrlToPathWithRelativeUrls()
+    {
+        $reflection = new \ReflectionClass(FilesystemPublisher::class);
+        $urlToPath = $reflection->getMethod('URLtoPath');
+        $urlToPath->setAccessible(true);
+
+        $this->assertEquals(
+            'index',
+            $urlToPath->invokeArgs($this->fsp, ['/'])
+        );
+
+        $this->assertEquals(
+            'about-us',
+            $urlToPath->invokeArgs($this->fsp, ['about-us'])
+        );
+
+        $this->assertEquals(
+            'parent/child',
+            $urlToPath->invokeArgs($this->fsp, ['parent/child'])
+        );
+    }
+
+    public function testUrlToPathWithAbsoluteUrls()
+    {
+        $reflection = new \ReflectionClass(FilesystemPublisher::class);
+        $urlToPath = $reflection->getMethod('URLtoPath');
+        $urlToPath->setAccessible(true);
```

Comment:
This opens up the debate about whether we should test protected methods: https://stackoverflow.com/questions/249664/best-practices-to-test-protected-methods-with-phpunit

There's a few cases of this in this integration test class - perhaps we should just make the method public and be done with it?

Reply:
We only want to expose public methods for `StaticPublisher` as that's the interface we are implementing. `URLtoPath` is very much an internal method for the `FilesystemPublisher` which doesn't make sense in the context of an abstract "publisher" engine.

Making it public will mean people make assumptions about it being available on all Publishers and that's not accurate.

We want to test the internal logic of this class whilst not exposing the methods for all to see. Whilst I can relate to the argument that we should test the methods that rely on this method, that provides implicit coverage whereas we want to provide some explicit coverage of this method.

The alternative is just to remove the tests IMO